gems/aws-crt/lib/aws-crt/native.rb

```diff
@@ -10,30 +9,77 @@ module Native
 
       ffi_lib [crt_bin_path(local_platform), 'libaws-crt']
 
+      # aws_byte_cursor binding
+      class ByteCursor < FFI::Struct
+        layout :len, :size_t,
+               :ptr, :pointer
+
+        def to_s
+          return unless (self[:len]).positive? && !(self[:ptr]).null?
+
+          self[:ptr].get_string(0, self[:len])
+        end
+      end
+
       # Core API
       attach_function :init, :aws_crt_init, [], :void
       attach_function :last_error, :aws_crt_last_error, [], :int
       attach_function :error_str, :aws_crt_error_str, [:int], :string
       attach_function :error_name, :aws_crt_error_name, [:int], :string
       attach_function :error_debug_str, :aws_crt_error_debug_str, [:int], :string
       attach_function :reset_error, :aws_crt_reset_error, [], :void
-      attach_function :global_thread_creator_shutdown_wait_for, :aws_crt_global_thread_creator_shutdown_wait_for, [:uint32], :int
+
+      # This MUST follow definitions of core error functions since it relies on them
+      # And error functions should NOT be wrapped.
+      #
+      # Overridden for three purposes.
+      #
+      # 1. Allows us to only supply the aws_crt C name, and converts it removes
+      #     the aws_crt.
+      # 2. Wraps the call in an error-raise checker.
+      # 3. Creates a bang method that does not do automatic error checking.
+      def self.attach_function(c_name, params, returns, options = {})
```

Comment:
debatable: Rather than overwrite the existing `attach_function`, and then you need to call it with bang to skip error checking... maybe we should write explicit variants of `attach_function, like:

```
attach_nothrow_fn :aws_crt_last_error, [], :int
attach_ptrerr_fn :aws_crt_event_loop_group_new, [:uint16], :pointer
attach_interr_fn :aws_crt_credentials_get_access_key_id, [:pointer], ByteCursor.by_value
```

Reply:
Yeah - I think that probably makes sense - rather than defining some functions before overriding the `attach_function`.  Probably don't need seperate ptrerr and interr versions though.  I'll make that change.